packages/plugin-inline/Inline.ts

```diff
@@ -0,0 +1,106 @@
+     * in the next insert which format should be used.
+     */
+    toggleFormat(params: FormatParams): void {
+        const format = params.format;
+        const range = this.editor.vDocument.selection.range;
+        if (range.isCollapsed()) {
+            if (!this.formatCache) {
+                this.formatCache = this.getCurrentFormats();
+            }
+            const index = this.formatCache.findIndex(f => f.name === format.name);
+            if (index !== -1) {
+                this.formatCache.splice(index, 1);
+            } else {
+                this.formatCache.push(format.shallowDuplicate());
+            }
+        } else {
+            const selectedInlines = range.selectedNodes(InlineNode);
+
+            // If every char in the range has the format `format.name`, remove
+            // the format for all of them.
+            const allHaveFormat = selectedInlines.every(inline => {
+                return !!inline.formats.find(f => f.name === format.name);
+            });
+            if (allHaveFormat) {
+                selectedInlines.forEach(inline => {
+                    const index = inline.formats.findIndex(f => f.name === format.name);
+                    const matchingFormat = inline.formats[index];
+                    if (matchingFormat.attributes && matchingFormat.attributes.size) {
+                        matchingFormat.attributes.forEach((value: string, key: string) => {
+                            if (!inline.attributes) {
+                                inline.attributes = new Map<string, string>();
+                            }
+                            inline.attributes.set(key, value);
+                        });
+                    }
+                    inline.formats.splice(index, 1);
+                });
+                // If there is at least one char in the range without the format
+                // `format.name`, set the format for all nodes.
+            } else {
+                selectedInlines.forEach(inline => {
+                    if (!inline.formats.find(f => f.isSameAs(format))) {
+                        inline.formats.push(format.shallowDuplicate());
+                    }
+                });
+            }
+        }
+    }
+    /**
+     * Get the format for the next insertion.
+     */
+    getCurrentFormats(range = this.editor.vDocument.selection.range): Format[] {
+        let formats: Format[] = [];
+        if (this.formatCache) {
+            return this.formatCache;
+        } else if (range.isCollapsed()) {
+            const inlineToCopyFormat = range.start.previousSibling() || range.start.nextSibling();
+            if (inlineToCopyFormat && inlineToCopyFormat.is(InlineNode)) {
+                formats = [...inlineToCopyFormat.formats];
+            }
+        } else {
+            const selectedInlines = range.selectedNodes(InlineNode);
+            selectedInlines.forEach(inline => (formats = [...formats, ...inline.formats]));
```

Comment:
I'm not sure this is right.
Consider this:
`<b>ab[cdef</b>ghijkl<i>mnopq</i>rstu<u>vwx]yz</u>`

And now you insert 'k'.

If I understood the code correctly, it's going to take all the formats of the inlines in the selection, that is `b`, `i,` and `u`. I *think* I would have expected to only write in `b` since `range.start` is only affected by bold.

Reply:
We only need take the format of the first char in the range.

Reply:
```typescript
/**
 * Get the format for the next insertion.
 */
getCurrentFormats(range = this.editor.vDocument.selection.range): Format[] {
    let formats: Format[] = [];
    if (this.formatCache) {
        return this.formatCache;
    } else {
        // If the range has a previous sibling, take its formats.
        const previousSibling = range.start.previousSibling();
        if (previousSibling && previousSibling.is(InlineNode)) {
            formats = [...previousSibling.formats];
        }
    }
    return formats;
}
```

Reply:
So actually this is right because it is the current format. It's in `insertText` that we need to make sure to take the current format _after_ emptying the range.

Reply:
I think I would change the `else` clause of `getCurrentFormats` to:
```typescript
const inlineToCopyFormat = range.start.nextSibling();
if (inlineToCopyFormat && inlineToCopyFormat.is(InlineNode)) {
    formats = [...inlineToCopyFormat.formats];
}
```

I think it would solve the issue I was pointing and would work for your test as well.

Added bonus: the `if` and `else` clauses would become so similar that the entire function could be simplified to:
```typescript
if (this.formatCache) {
    return this.formatCache;
}

let inlineToCopyFormat: VNode;
if (range.isCollapsed()) {
    inlineToCopyFormat = range.start.previousSibling() || range.start.nextSibling();
} else {
    inlineToCopyFormat = range.start.nextSibling();
}
if (inlineToCopyFormat && inlineToCopyFormat.is(InlineNode)) {
    return [...inlineToCopyFormat.formats];
}

return [];
```

And at this point I'm not even sure we would need a format cache since computing this would not be so expensive (but we still want inserting text to be quite fast so we probably somewhat fear to remove it, though.)